feat: Add expiration date to access token & hmac keys (#1219)
codeigniter4/shield: docs/references/authentication/hmac.md

$user->revokeAllHmacTokens();
```

## Expiring HMAC Keys

By default, the HMAC keys don't expire unless they meet the HMAC Keys lifetime expiration after their last used date.

HMAC keys can be set to expire through the `generateHmacToken()` method. This takes the expiration date as the $expiresAt argument. It's also possible to update an existing HMAC key using `setHmacTokenExpirationById($HmacTokenID, $expiresAt)`

`$expiresAt` [Time](/libraries/time.html) object

```php
// Expiration date = 2024-11-03 12:00:00
$expiresAt = Time::parse('2024-11-03 12:00:00');

> $token = $this->user->generateHmacToken('foo', ['foo:bar'], $expiresAt);

Reviewer: ```suggestion
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);
```
Scope names should avoid using a colon, according to the docs.

// Expiration date = 2024-11-15 00:00:00
$expiresAt = Time::parse('2024-11-15 00:00:00');
$token = $user->setHmacTokenExpirationById($token->id, $expiresAt);

// Or Expiration date = 1 month + 15 days into the future

> $expiresAt = Time::now();
> $expiresAt = $expiresAt->addMonths(1);
> $expiresAt = $expiresAt->addDays(15);

Reviewer: ```suggestion
$expiresAt = Time::now()->addMonths(1)->addDays(15);
```

$token = $user->setHmacTokenExpirationById($token->id, $expiresAt);
```

The following support methods are also available:

> `hasHmacTokenExpired(AccessToken $HmacToken)` - Checks if the HMAC key has expired. Returns `true` if the HMAC key has expired, `false` if not, and `null` if the expire date is null.

Reviewer: ```suggestion
`hasHmacTokenExpired(AccessToken $HmacToken)` - Checks if the HMAC key has expired. Returns `true` if the HMAC key has expired, `false` if not.
```

```php
$expiresAt = Time::parse('2024-11-03 12:00:00');

> $token = $this->user->generateHmacToken('foo', ['foo:bar'], $expiresAt);

Reviewer: ```suggestion
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);
```

$this->user->hasHmacTokenExpired($token); // Returns true
```

`canHmacTokenExpire(AccessToken $HmacToken)` - Checks if HMAC key has an expiration set. Returns `true` or `false` accordingly.

```php
$expiresAt = Time::parse('2024-11-03 12:00:00');

> $token = $this->user->generateHmacToken('foo', ['foo:bar'], $expiresAt);

Reviewer: ```suggestion
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);
```

$user->revokeAllHmacTokens();
```

## Expiring HMAC Keys

By default, the HMAC keys don't expire unless they meet the HMAC Keys lifetime expiration after their last used date.

> HMAC keys can be set to expire through the `generateHmacToken()` method. This takes the expiration date as the $expiresAt argument. To update or remove an existing HMAC key expiration date use `setHmacTokenExpirationById($HmacTokenID, $expiresAt)`

Reviewer 2: ```suggestion
HMAC keys can be set to expire through the `generateHmacToken()` method. This takes the expiration date as the `$expiresAt` argument. To update or remove an existing HMAC key expiration date use `setHmacTokenExpirationById($HmacTokenID, $expiresAt)`
```

> HMAC keys can be set to expire through the `generateHmacToken()` method. This takes the expiration date as the $expiresAt argument. To update or remove an existing HMAC key expiration date use `setHmacTokenExpirationById($HmacTokenID, $expiresAt)`

Reviewer 2: _To update or remove an existing HMAC key expiration date use `setHmacTokenExpirationById($HmacTokenID, $expiresAt)`_

Since these are two distinct responsibilities, I think it's better to create separate functions for updating and removing the HMAC token expiration. This improves code clarity and maintainability.  

Update expiration date:
   ```php
   public function updateHmacTokenExpiration($hmacTokenID,  $expiresAt): bool
   ```
Remove expiration date (set to null)
   ```php
   public function removeHmacTokenExpiration($hmacTokenID): bool
   ```

What do you think?

Reviewer 3: Yes, I also think `update***()` and `remove***()` would be better.

Author: Yeah, michalsn and I talked about that same question (separation) earlier and kept everything on the same method.

However, I understand that the difference in functionality should warrant their own methods, also makes the removal of the expiration more conscious and less error prone action. I agree in splitting the actions.

> `$expiresAt` [Time](/libraries/time.html) object

Reviewer 2: ```suggestion
`$expiresAt` [Time](https://codeigniter.com/user_guide/libraries/time.html) object
```

```php
// Expiration date = 2024-11-03 12:00:00
$expiresAt = Time::parse('2024-11-03 12:00:00');

> $token = $this->user->generateHmacToken('foo', ['foo:bar'], $expiresAt);

Reviewer 2: ```suggestion
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);
```

// Expiration date = 2024-11-15 00:00:00
$expiresAt = Time::parse('2024-11-15 00:00:00');
$token = $user->setHmacTokenExpirationById($token->id, $expiresAt);

// Expiration date = 1 month + 15 days into the future
$expiresAt = Time::now()->addMonths(1)->addDays(15);
$token = $user->setHmacTokenExpirationById($token->id, $expiresAt);

// Remove the expiration date
$token = $user->setHmacTokenExpirationById($token->id, null);
```

The following support methods are also available:

> `hasHmacTokenExpired(AccessToken $HmacToken)` - Checks if the HMAC key has expired. Returns `true` if the HMAC key has expired, `false` if not.

Reviewer 2: Wouldn't it be better to use `isHmacTokenExpired` instead of `hasHmacTokenExpired`? 
This name is clearer and more naturally indicates that the function checks whether the token has expired.

Author: I agree it's better semantically.

```php
$expiresAt = Time::parse('2024-11-03 12:00:00');
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);

$this->user->hasHmacTokenExpired($token); // Returns true
```

> `canHmacTokenExpire(AccessToken $HmacToken)` - Checks if HMAC key has an expiration set. Returns `true` or `false` accordingly.

Reviewer 2: Wouldn't `hasHmacTokenExpiry` be a better choice instead of `canHmacTokenExpire`? 
The word **hasExpiry** more clearly indicates whether the token has an expiration set, avoiding potential confusion with can, which might imply permission or the ability to change.

```php
$expiresAt = Time::parse('2024-11-03 12:00:00');

$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);
$this->user->canHmacTokenExpire($token); // Returns true

$token2 = $this->user->generateHmacToken('bar');
$this->user->canHmacTokenExpire($token2); // Returns false
```

You can also easily set all existing HMAC keys/tokens as expired with the `spark` command:

> ```

Reviewer 2: ```suggestion
```console
```

php spark shield:hmac invalidateAll
```

> **Careful!** This command invalidates _all_ keys for _all_ users.

Reviewer 2: ```suggestion
**Careful!** This command invalidates _all_ keys for _all_ users.

!!! warning

    This command invalidates _all_ keys for _all_ users.
```

`$expiresAt` [Time](https://codeigniter.com/user_guide/libraries/time.html) object

```php
// Expiration date = 2024-11-03 12:00:00
$expiresAt = Time::parse('2024-11-03 12:00:00');
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);

// Expiration date = 2024-11-15 00:00:00
$expiresAt = Time::parse('2024-11-15 00:00:00');
$user->updateHmacTokenExpiration($token->id, $expiresAt);

// Expiration date = 1 month + 15 days into the future
$expiresAt = Time::now()->addMonths(1)->addDays(15);
$user->updateHmacTokenExpiration($token->id, $expiresAt);

// Remove the expiration date
$user->removeHmacTokenExpiration($token->id);
```

The following support methods are also available:

`isHmacTokenExpired(AccessToken $HmacToken)` - Checks if the HMAC key is expired. Returns `true` if the HMAC key is expired, `false` if not.

```php
$expiresAt = Time::parse('2024-11-03 12:00:00');
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);

> $this->user->hasHmacTokenExpired($token); // Returns true

Reviewer 2: Did you mean?
```suggestion
$this->user->isHmacTokenExpired($token); // Returns true
```

$user->revokeAllHmacTokens();
```

## Expiring HMAC Keys

> By default, the HMAC keys don't expire unless they meet the HMAC Keys lifetime expiration after their last used date.

Reviewer 4: ```suggestion
By default, the HMAC keys don't expire unless they reach the HMAC keys' lifetime expiration after their last use date.
```

> HMAC keys can be set to expire through the `generateHmacToken()` method. This takes the expiration date as the `$expiresAt` argument. To update an existing HMAC key expiration date use `updateHmacTokenExpiration($hmacTokenID, $expiresAt)` and to remove `removeHmacTokenExpiration($hmacTokenID)`.

Reviewer 4: ```suggestion
HMAC keys can be set to expire through the `generateHmacToken()` method. This takes the expiration date as the `$expiresAt` argument. To update an existing HMAC key expiration date, use `updateHmacTokenExpiration($hmacTokenID, $expiresAt)`. To remove it, use `removeHmacTokenExpiration($hmacTokenID)`.
```

`$expiresAt` [Time](https://codeigniter.com/user_guide/libraries/time.html) object

```php
// Expiration date = 2024-11-03 12:00:00
$expiresAt = Time::parse('2024-11-03 12:00:00');
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);

// Expiration date = 2024-11-15 00:00:00
$expiresAt = Time::parse('2024-11-15 00:00:00');
$user->updateHmacTokenExpiration($token->id, $expiresAt);

// Expiration date = 1 month + 15 days into the future
$expiresAt = Time::now()->addMonths(1)->addDays(15);
$user->updateHmacTokenExpiration($token->id, $expiresAt);

// Remove the expiration date
$user->removeHmacTokenExpiration($token->id);
```

The following support methods are also available:

> `isHmacTokenExpired(AccessToken $HmacToken)` - Checks if the HMAC key is expired. Returns `true` if the HMAC key is expired, `false` if not.

Reviewer 4: ```suggestion
`isHmacTokenExpired(AccessToken $hmacToken)` - Checks if the HMAC key is expired. Returns `true` if the HMAC key is expired; otherwise, returns `false`.
```

```php
$expiresAt = Time::parse('2024-11-03 12:00:00');
$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);

$this->user->isHmacTokenExpired($token); // Returns true
```

> `hasHmacTokenExpiry(AccessToken $HmacToken)` - Checks if HMAC key has an expiration set. Returns `true` or `false` accordingly.

Reviewer 4: ```suggestion
`hasHmacTokenExpiry(AccessToken $hmacToken)` - Checks if HMAC key has an expiration set. Returns `true` if the HMAC key is expired; otherwise, returns `false`.
```

```php
$expiresAt = Time::parse('2024-11-03 12:00:00');

$token = $this->user->generateHmacToken('foo', ['foo.bar'], $expiresAt);

> $this->user->canHmacTokenExpire($token); // Returns true

Reviewer 2: And ?

$token2 = $this->user->generateHmacToken('bar');

> $this->user->canHmacTokenExpire($token2); // Returns false

Reviewer 2: And?

```

You can also easily set all existing HMAC keys/tokens as expired with the `spark` command:
```console
php spark shield:hmac invalidateAll
```

!!! warning

>     This command invalidates _all_ keys for _all_ users.

Reviewer 2: ```suggestion

    This command invalidates _all_ keys for _all_ users.
```

public $unusedTokenLifetime = YEAR;
```

> ### HMAC Keys Expiration vs Lifetime
> Expiration and Lifetime are different concepts. The lifetime is the maximum time allowed for the HMAC Key to exist since its last use. HMAC Key expiration, on the other hand, is a set date in which the HMAC Key will cease to function.

Reviewer: ```suggestion
### HMAC Keys Expiration vs Lifetime

Expiration and Lifetime are different concepts. The lifetime is the maximum time allowed for the HMAC Key to exist since its last use. HMAC Key expiration, on the other hand, is a set date in which the HMAC Key will cease to function.
```
Just formatting - one line space between the heading and the content.

public $unusedTokenLifetime = YEAR;
```

### HMAC Keys Expiration vs Lifetime

> Expiration and Lifetime are different concepts. The lifetime is the maximum time allowed for the HMAC Key to exist since its last use. HMAC Key expiration, on the other hand, is a set date in which the HMAC Key will cease to function.

Reviewer 4: ```suggestion
Expiration and lifetime are two different concepts. The lifetime is the maximum time allowed for the HMAC Key to exist since its last use. HMAC Key expiration, on the other hand, is a set date in which the HMAC Key will cease to function.
```
